src/main/java/com/amazon/ion/IonEncodingVersion.java

```diff
@@ -15,30 +18,39 @@
  *
  * @param <BinaryWriterBuilder> the type of binary writer builder compatible with this version.
  */
-// TODO add a parameter for the text writer builder type; add a "textWriterBuilder()" method.
-public abstract class IonEncodingVersion<BinaryWriterBuilder> {
+public abstract class IonEncodingVersion<BinaryWriterBuilder, TextWriterBuilder> {
```

Comment:
FWIW, this feels like its treading into `IonSystem` territory, and that's a facade class I deeply regret, since it couples too many things together.   The replacement builder-pattern was intended to avoid coupling across different features, to ease things like tree-shaking and partitioning the library into smaller portions.

A simpler approach, IMO, would be to add things like `IonBinaryWriterBuilder.standard11()`.

Reply:
@toddjonker, here's the line of reasoning:

Regarding version-specific sub interfaces... we are trying to work within the constraints of what already exists in `ion-java`, and at the same time, we are trying to get away from all of the behavioral inheritance (`extends`) and try to just use API inheritance (`implements`).

All of the interfaces are here to support various use cases. I think we need to have a general `IonWriterBuilder` interface, as well as an interface for Ion 1.0 builders and Ion 1.1 builders, and interfaces for text builders and binary builders. That way, you can write code that e.g. expects an Ion 1.1 builder but is generic over whether it's text or binary, or e.g. expects a text builder but is agnostic to whether it is Ion 1.0 or 1.1. Finally, we need to have the version/encoding specific interfaces so that someone can get a builder that has exactly the options that are relevant to the Ion encoding that they desire.

It does feel like a lot of different interfaces, but I don't think it's unreasonable. What makes it seem unreasonable is the way we've had to shove it into the existing abstract-class inheritance structure of the builders.

As for the `IonEncodingVersion` class:

We considered having a `withMinorVersion(int version)` option on the builder. However, that was flawed because
* we would have to validate whether the specified version is valid when we could just use the type system to prevent someone from passing an invalid version
* `withMinorVersion` will become awkward if there's ever a new major version of Ion.
* it always returns the same type of (version-agnostic) builder

So, we considered having a `withIonVersion(IonEncodingVersion version)` where `IonEncodingVersion` is an enum. That solves some of the issues, but also is problematic because adding a new enum variant is a potentially breaking change for languages (looking at you Kotlin) that try to enforce exhaustive matches on enums.

So then we considered a "pseudo" enum—i.e. a class with a private constructor and some public constants that are effectively the enum variants.

Finally, we realized that we could go one step further and just use the `IonEncodingVersion` instances as the entry point to version-specific builder APIs.

If this pattern gets negative feedback from our customers, we can create methods such as `standard11()`, but we've been trying to avoid that so far because `11` looks like `11` rather than `1.1`, and adding underscores, i.e. `standard_1_1()`, looks very _not_-Java.

We could also create methods such as this (but probably in Java):
```kotlin
fun <TextBuilder> standardText(version: IonVersion<TextBuilder, *>): TextBuilder
fun <BinaryBuilder> standardBinary(version: IonVersion<*, BinaryBuilder>): TextBuilder
```

**TLDR; `IonEncodingVersion` is just a way to represent the Ion version as an object/type rather than as an integer (or pair of integers). It can be a static entry point to _builders_, but should never become more than that.**